### PR TITLE
Allow a device to be assigned to application

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

Permit a device to be assigned to an application.

NOTE: Raised in draft to permit other engineers to test - before writing a whole bunch of tests and updating docs.

Notable:
1. permit the agent to start even if the working dir does not exist. Instead of a hard exit, an attempt to generate the working directory is made.
2. adds `currentApplication` which when set, permits the device to launch node-red - allowing user to modifiy flows.
3. updates the log output to indicate instance or application as device owner
4. only writes necessary parts to file - to prevent the fake application snapshot overwriting users flows edits

#### Application level device implementation:
 - Developer Mode / Remote Access ✅
 - Device Logs ✅
 - Team Library ✅
 - Theme, device name and NR menu links ✅
 - Adding device in application view adds device to application ✅
 - Newly added device in application view updates status when config is applied to device and it starts up ✅
 - Upload Snapshot ❌
 - Download Snapshot ❌
 - Project Links ❌
 - Grouping and settings ❌
- Module cache for device at application level  ❌

#### TODO (this iteration):
 - [x] tests
 - [x] docs
 - [ ] manual testing combinations - to ensure there are no issues when running older platform with newer agent and vice versa
  
#### TODO (next iteration)
 - Provisioning Tokens for auto add to application
 - Module cache for device at application level
 - Support snapshots (upload/download)
 - Support grouping
 - Group settings (env/modules/etc)
 - Project Links (Device links? Concept?)

## Related Issue(s)

[Devices to be assigned to an Application, without being assigned to an Instance#2619](https://github.com/flowforge/flowforge/issues/2619)

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Review by knolleary — CHANGES_REQUESTED** [review]
Some initial comments... not got through everything yet.

**Comment by knolleary** [comment]
In relation to the discussion on https://github.com/flowforge/flowforge/pull/2621#discussion_r1304446272 , I can see a possible future where the valid states will be.

Assigned To | Application | Instance
-------------|------------|----------
_nothing_     | _null_         | _null_
Application | `<applicationId>`          | _null_
Instance   | `<applicationId>`             | `<instanceId>`

Meaning, the presence of `application` means its assigned to _something_, and `instance` tells you if it is a specific instance or not.

In quite a few places with this PR there is pieces of logic similar to:

```
(this.currentApplication ? 'application' : (this.currentProject ? 'project' : 'none'))
```

Which puts all the weight on whether application is set or not. That matches up with the current implementation of the platform, but would make it harder **if** we decide to modify the model as I've described.

I *think* we could safely flip the logic in the device agent to something like:

```
this.currentProject ? ' project' : (this.currentApplication ? 'application' : 'none'))
```

I believe that would still behave the same way with the current server implementation, but would also work *if* we modified things.

Given how many places that bit of logic exists in this PR, I don't know if the logic is sound - but wanted to get your input if this is feasible?

File: index.js [hunk 1]
```diff
@@ -41,10 +41,18 @@ function main (testOptions) {
 
     // Require dir to be created
     if (!fs.existsSync(options.dir)) {
-        const quitMsg = `Cannot find dir '${options.dir}'.
-    Please ensure it exists and is writable, or set a different path with -d`
-        quit(quitMsg, 20) // Exit Code 20 - Invalid dir
-        // REF: https://slg.ddnss.de/list-of-common-exit-codes-for-gnu-linux/
+        try {
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply index.js:44]: When developing this, I needed to spin up multipe devices on demand and added this to ease it. 

I forgot I left it in but now it is - I would really prefer to leave it in as it is both user useful and much more user friendly than an error message and failed start.

I will update the docs to reflect this - but I can just as easily remove it if insisted upon.

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply index.js:44]: docs updated in [c849697](https://github.com/flowforge/flowforge-device-agent/pull/161/commits/c84969721d1ee0c4c9d4b1457cb598f54fa4b9df)

**Comment by Steve-Mcl** [comment]
> Looking at that line of code above, it puts all the weight on whether application is set or not. That matches up with the current implementation of the platform, but would make it harder **if** we decide to modify the model as I've described.
> 
> I _think_ we could safely flip the logic in the device agent to something like:
> 
> ```
> this.currentProject ? ' project' : (this.currentApplication ? 'application' : 'none'))
> ```
> 
> I believe that would still behave the same way with the current server implementation, but would also work _if_ we modified things.
> 
> Given how many places that bit of logic exists in this PR, I don't know if the logic is sound - but wanted to get your input if this is feasible?

The order of the logic was deliberate and considered but I also understand your point.

I will go through both ends again and determine the side-effects of changing (if any).

Watch this space.

**Comment by Steve-Mcl** [comment]
> > I _think_ we could safely flip the logic in the device agent to something like:
> > ```
> > this.currentProject ? ' project' : (this.currentApplication ? 'application' : 'none'))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I believe that would still behave the same way with the current server implementation, but would also work _if_ we modified things.
> > Given how many places that bit of logic exists in this PR, I don't know if the logic is sound - but wanted to get your input if this is feasible?
> 
> The order of the logic was deliberate and considered but I also understand your point.
> 
> I will go through both ends again and determine the side-effects of changing (if any).

Login order changed in [1026df4](https://github.com/flowforge/flowforge-device-agent/pull/161/commits/1026df47c6403bed7033a435cadd158e5e36e57c) and verified locally.

It did throw up an issue which has also been resolved on the platform repo.

Ready for re-review now.

**Review by knolleary — APPROVED** [review]

**Pull request merged**